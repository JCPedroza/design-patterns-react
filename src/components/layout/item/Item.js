import { arrayOf, object, string } from 'prop-types'

const buildItemToShow = (item, keysToShow) => {
  const itemProxy = new Proxy(item, { ownKeys: () => keysToShow })
  return Object.entries(Object.assign({}, itemProxy))
}

const Item = ({ item, keysToShow }) => {
  const itemToShow = buildItemToShow(item, keysToShow)

  return (
    <>
      <ul>
        {itemToShow.map(([key, value]) => <li key={key}>{key}: {value}</li>)}
      </ul>
    </>
  )
}

Item.propTypes = {
  item: object.isRequired,
  keysToShow: arrayOf(string).isRequired
}

export default Item
