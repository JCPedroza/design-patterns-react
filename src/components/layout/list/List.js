import { Fragment } from 'react'
import { arrayOf, object, string, func } from 'prop-types'

const List = ({
  items,
  keysToShow,
  resourceName,
  itemComponent: ItemComponent
}) => {
  return (
    <>
      {items.map((item, index) => (
        <Fragment key={item.id}>
          <h3>{index + 1}</h3>
          <ItemComponent
            key={item.id}
            {...{ [resourceName]: item }}
            keysToShow={keysToShow}
          />
        </Fragment>
      ))}
    </>
  )
}

List.propTypes = {
  items: arrayOf(object).isRequired,
  keysToShow: arrayOf(string).isRequired,
  resourceName: string.isRequired,
  itemComponent: func.isRequired
}

export default List
