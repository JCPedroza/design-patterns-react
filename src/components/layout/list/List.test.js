import { render } from '@testing-library/react'

import Item from '../item/Item'
import List from './List'

const items = [
  { id: 14, title: 'One', author: 'Mario', year: 1926, time: 123 },
  { id: 38, title: 'Tuesday', author: 'Fernanda', year: 2013, time: 234 },
  { id: 49, title: 'Untitled', author: 'Radiohead', year: 2003, time: 357 }
]

const keysToShow = ['id', 'title', 'author', 'year']
const resourceName = 'item'

it('renders a list', () => {
  const { container } = render(
    <List
      items={items}
      keysToShow={keysToShow}
      resourceName={resourceName}
      itemComponent={Item}
    />
  )

  expect(container).toMatchSnapshot()
})
