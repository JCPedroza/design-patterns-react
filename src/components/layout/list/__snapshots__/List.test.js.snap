// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list 1`] = `
<div>
  <h3>
    1
  </h3>
  <ul>
    <li>
      id
      : 
      14
    </li>
    <li>
      title
      : 
      One
    </li>
    <li>
      author
      : 
      Mario
    </li>
    <li>
      year
      : 
      1926
    </li>
  </ul>
  <h3>
    2
  </h3>
  <ul>
    <li>
      id
      : 
      38
    </li>
    <li>
      title
      : 
      Tuesday
    </li>
    <li>
      author
      : 
      Fernanda
    </li>
    <li>
      year
      : 
      2013
    </li>
  </ul>
  <h3>
    3
  </h3>
  <ul>
    <li>
      id
      : 
      49
    </li>
    <li>
      title
      : 
      Untitled
    </li>
    <li>
      author
      : 
      Radiohead
    </li>
    <li>
      year
      : 
      2003
    </li>
  </ul>
</div>
`;
